sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusSenderClient.cs

```diff
@@ -0,0 +1,499 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Messaging.ServiceBus.Core;
+using Azure.Messaging.ServiceBus.Diagnostics;
+using Microsoft.Azure.Amqp;
+
+namespace Azure.Messaging.ServiceBus.Sender
+{
+    /// <summary>
+    ///   A client responsible for sending <see cref="ServiceBusMessage" /> to a specific Service Bus entity (queue or topic).
+    /// </summary>
+    ///
+    public class ServiceBusSenderClient : IAsyncDisposable
```

Comment:
This comes from our .NET guidelines for service client naming conventions - https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-client

✅ DO name service client types with the Client suffix.

Reply:
That guidance is already off because there's no single client for the service. There are Queue, Topic, and Subscription clients. It's my personal dissatisfaction with the name as it sounds as buttery butter just to keep the checkbox ticked off.

Reply:
I don't think the guidance applies only when there is a single client, e.g. Storage has several clients all of which are suffixed with `Client`.

Reply:
Indeed. You can see the parallels between Storage Blob / Queue / Table and Service Bus Queue / Topic / Subscription. Message Sender and Receiver are very different as they span all 3 entity types. If there was only one, I'd be inclined to call it `ServiceBusClient`. But they are not 🙂

Reply:
We haven't solidified on the client hierarchy yet. We may end up having separate clients for each entity type. The working prototype here is to have separate clients per entity type for receiver, but one client for sending to any entity. The `ServiceBusReceiverClient` is an abstract class from which the other entity receivers derive. In Storage there is also a `BlobBaseClient` that can act on different blob types (other clients derive from it for specific blob types), which seems analogous to having a SenderClient that can send to different entities.

Reply:
> The working prototype here is to have separate clients per entity type for receiver, but one client for sending to any entity.

I'll just note that having a type that can receive from any entity type is crucial. Customers don't always create an entity-specific type. Sometimes it's a single type that can process messages from either a queue or a subscription.

Reply:
Hmm, do you mean that customers might not know what type of entity it is that they will be receiving from but they will know the name of it? We need to specify the entity name in the constructor, so they would only be receiving from a particular entity from that client. We could make `ServiceBusReceiverClient` not be abstract so it can be instantiated directly, but I'm trying to understand the scenario where you wouldn't know the entity type.

Reply:
This is why learning from Track1 and Track0 authors would be vital.
The library is used in many ways. When you want to receive messages w/o differentiating what entity it's coming from, you have an entity path and that's it. It could be a queue name or a topic+subscription name that would provide the path. That is why `MessageReceiver` is so great. Similar to `MessageSender`, same idea. You don't care if you send to a topic or a queue, you just send it to whatever the path there is.
/cc @nemakam

Reply:
We absolutely want to leverage the knowledge of Track 1 and Track 0 authors :)

For the sending side, we currently only have one client which could send to any entity. I think we could make `ServiceBusReceiverClient` not be abstract to fulfill this same purpose, but the main reason I was hesitant about doing this is that the `QueueReceiverClient` is essentially identical to `ServiceBusReceiverClient` so it seemed odd to provide both clients. However, I see how it might occur where an app just has a connection string with an entityName in it and they don't need to know anything more than that to start consuming.

Reply:
I'm interested to hear your thoughts on the dichotomy of having the entity specific clients only on the receiver side, and just one generic client on the sender side.

Reply:
The original thinking in Track1 was that entity-specific clients are for more simple scenarios/use cases. Message Sender and Receiver would be for more advanced scenarios. That has definitely was the case in my experience. Folks that wanted simplicity while working with a specific entity type could easily send and receive messages. Receiving messages was simplified to a message handler that would only require the user callback and "magic" would happen on their behalf if they want to perform concurrent processing. At the same time, for power users, there were Sender and Receiver that could allow building more sophisticated/custom message pumps. This is BTW the route we've chosen. So it was a very clear cut - `QueueClient`/`TopicClient`/`SubscriptionClient` for simple scenarios and `MessageSender`+`MessageReceiver` for advanced scenarios.

> I'm interested to hear your thoughts on the dichotomy of having the entity specific clients only on the receiver side, and just one generic client on the sender side.

If I understand correctly what you're asking, you're going to position it in a way that in order to perform a receive operation one must know what type of entity they'll be receiving from. In scenarios where the topology is known upfront, it might be OK. For scenarios where an entity is not necessarily known aside from the full path, this would be painful. E.g. frameworks built on top of ASB SDK. For a given entity path, the code would require an additional piece of information to identify the type of entity to leverage the correct client. That's a deal-breaker as suddenly a new piece of information would need to be reinvented and passed around.

In case I misunderstood the question, please elaborate.

Reply:
Sorry, I should have clarified that I was operating under the assumption that we will most likely want to have the generic receiver based on your feedback. I am wondering if it would seem strange to only have entity-specific clients on the receiver side (in addition to a session receiver client), while on the sender side simply having one generic client. Entity specific receivers make sense since subscriptions have additional operations that don't apply to queues, as do sessions. But for message sending, there doesn't appear to be a need for entity level senders.

Reply:
It would be weird for this operating on a single entity level such as queue as they'd need to use two constructs, one to receive and another one to send messages. That, I'd even dare to say, goes against the new Azure SDKs mantra of having it simple for customers 🙂

Reply:
The reason we thought this would work is that we couldn't come up with a common use case where someone would need to both send and receive from the same entity in the context of one application (one might receive from one entity, and send to a different entity, but that would already require having two separate clients). We expect there to be a receiving application (or part of the application) and a separate sending one. Working with this assumption, it seemed sensible to split by send/receive clients.
Would be happy to hear examples of where this line of thinking doesn't hold up 😃 
@nemakam in case he has thoughts on this

Reply:
> we couldn't come up with a common use case where someone would need to both send and receive from the same entity in the context of one application (one might receive from one entity, and send to a different entity, but that would already require having two separate clients).

The dangers of the wors ass-u-me 😃 
I have a use case just like that that is quite common. When an endpoint is sending a message (local command) to itself, it will receive it from its input queue where it will dispatch it to earlier.

Reply:
Ha yes this is still very preliminary so please don't take any initial prototypes as a final decision on what the product will be. We expect to iterate on the design based on feedback from contributors like yourself! 

In the use case that you mention, what would be the reason for sending the message to itself? It sounds sort of like deferring a message, but what would have necessitated the client sending the message in the first place if it will be receiving it itself?

Reply:
Not deferring. Separation of concerns. If you have two handlers within the same logical boundary X and Y; when the handler X processes an incoming message (or does something else) and wants to instruct handler Y to perform a different operation without the temporal coupling of calling Y, a local command can be issued. Another use case is scheduling a message to self in the future, **not deferring**. Deferring in SB is very different.

Reply:
That's fair. Of course we could still have Handler X have access to both a receiver and sender client in this case.